### PR TITLE
Recipe app

### PR DESCRIPTION
**Pull request by Chandan-devs-tech**
# Description :
The Recipe app keeps track of all your recipes, ingredients, and inventory. It will allow you to save ingredients, keep track of what you have, create recipes, and generate a shopping list based on what you have and what you are missing from a recipe. Also, since sharing recipes is an important part of cooking the app should allow you to make them public so anyone can access them.
- The app will have some common interfaces. These are some of the interfaces and some wireframes for some of them:
  - A login page.
  - A registration page.
  - A food list (with all CRUD implementation, except for 'update').
  - A recipes list (with all CRUD implementation, except for 'update').
  - A general shopping list view (all missing food for all your recipes and total price).
  - A list of all public recipes from other users with their names and total prices.
## Project requirements
- You should follow the layout of the wireframes provided. You should personalize the rest of the design including colors, typographies, spacings, etc.
- Login page and registration page:
  - Should be built with Devise.
- Food list (only for a group that has less than 3 members):
  - Should display a list of food added by the logged-in user as in the wireframe (display also quantity of a given food).
  - Should lead to a form that allows users to add new food.
- Recipes list:
Should display a list of recipes created by the logged-in user as in the wireframe.
  - Should lead to recipe details.
  - If the user is the owner of the recipe, should allow the user to delete it.
- Public recipe list:
  - Should display a list of all public recipes ordered by newest as in the wireframe.
  - Should lead to recipe details.
- Recipe details:
  - Should display a toggle button that allows for a recipe to be made public or private.
  - If the recipe is public or the user is the owner of the recipe, should display the recipe details as in the wireframe.
  - If the user is the owner of the recipe, should lead to the form that allows the user to add new food.
- General shopping list view (only for a group that has less than 3 members):
  - Should show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
  - Should count the total food items and total price of the missing food.
- Make sure there are no N+1 queries happening.
- Create a navigation menu that allows users to open all of the pages you created.
- Write unit and integration tests
- Technical setup
  - Set up the repository on GitHub and use Gitflow.
  - Set up Devise for authentication.
  - Set up RSpec and Capybara testing libraries.
  
Imp

### CODE REVIEW

File: app/views/foods/show.html.erb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<p style="color: green"><%= notice %></p>
+
+<%= render @food %>
+
+<div>
+  <%#= link_to "Edit this food", edit_food_path(@food) %> |
+  <%= link_to "Back to foods", foods_path %>
+
+  <%#= button_to "Destroy this food", @food, method: :delete %>
```
Comment (rloterh) [inline app/views/foods/show.html.erb]: - [ ] [OPTIONAL]
There are commented code blocks found in your code on the repository. Kindly endeavor to remove **all** commented code from your code before pushing it to the repository so as to keep your code base well organized.

Comment (rloterh) [inline app/views/foods/show.html.erb]: - [ ] Nice job so far 👌, however, let's ensure that the foods table matches the given [wireframe](https://github.com/microverseinc/curriculum-rails/blob/main/recipe-app/images/food_index.png?raw=true) as much as possible; The `Add Food` or New Food button should be placed at the top right. Let's improve upon this aspect.

**Your design:**        |  **The original design:**
:-------------------------:|:-------------------------:
 | Home Page

File: app/views/recipes/_recipe.html.erb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      </tr>
+    <% end %>
+  </table>
+
+  <%# <p> %>
+    <%# <strong>User:</strong> %>
+    <%#= recipe.user_id %>
+  <%# </p> %>
+
+</div>
```
Comment (rloterh) [inline app/views/recipes/_recipe.html.erb]: - [ ] Let's also ensure a proper layout for a Recipe detail.

**Your design:**        |  **Expected:**
:-------------------------:|:-------------------------:
Home Page | Home Page

File: app/controllers/shopping_list_controller.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    # end
+    displayed_names = []
+    @recipe_foods.each do |recipe_food|
+      unless displayed_names.include?(recipe_food.food.name)
+        displayed_names << recipe_food.food.name
+        @total_price += recipe_food.food.price * recipe_food.quantity
+      end
+    end
+    @displayed_names = displayed_names
+  end
```
Comment (rloterh) [inline app/controllers/shopping_list_controller.rb]: - [ ] Let's ensure the general shopping list renders (In the data entered chicken is missing):
  * renders the list of food that is missing for all recipes of the logged-in user.
  * counts the total food items and total price of the missing food.
  
  **Your General shopping list view:**
  
image

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Chandan, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 
🚦 __🏃 🏡  🏚️ 🌴

## Highlights 👍  
- Authentication with Devise gem properly implemented!
- App schema closely follows given ERD!
- Properly addressed N+1 problem
- All users can view public recipes!
- Specs cases are passing!
- ReadMe looks awesome!

## Required Changes ♻️
_Kindly check the inline comments under the review._

## Optional suggestions 🧰 
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

_Kindly consider my inline comments marked **[OPTIONAL]** under this review_

**You can also consider:**
Let's make the ReadMe stand out even better by including instructions under the _Getting Started_ section, on how to run and test the app, for instance, using the commands  `rails server` and `rspec`.

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by OLIPLICHE — APPROVED** [review]
Hi @Chandan-devs-tech, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# Highlights
- Applied the changes requested by the reviewer (✅)) 👏
- The App is working fine 👏
- Commit messages are meaningful 👏 👏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**